// @superset-ui/core/esm/chart/models/ChartProps.js
import { createSelector } from 'reselect';
import { convertKeysToCamelCase } from '../..';
const DEFAULT_WIDTH = 800;
const DEFAULT_HEIGHT = 600;
export default class ChartProps {
  constructor(config = {}) {
    const {
      annotationData = {},
      datasource = {},
      formData = {},
      hooks = {},
      initialValues = {},
      queryData = {},
      queriesData = [],
      width = DEFAULT_WIDTH,
      height = DEFAULT_HEIGHT
    } = config;
    this.width = width;
    this.height = height;
    this.annotationData = annotationData;
    this.datasource = convertKeysToCamelCase(datasource);
    this.rawDatasource = datasource;
    this.formData = convertKeysToCamelCase(formData);
    this.rawFormData = formData;
    this.hooks = hooks;
    this.initialValues = initialValues;
    this.queryData = queryData;
    this.queriesData = queriesData;
  }

} // eslint-disable-next-line func-name-matching

ChartProps.createSelector = function create() {
  return createSelector(input => input.annotationData, input => input.datasource, input => input.formData, input => input.height, input => input.hooks, input => input.initialValues, input => input.queryData, input => input.queriesData, input => input.width, (annotationData, datasource, formData, height, hooks, initialValues, queryData, queriesData, width) => new ChartProps({
    annotationData,
    datasource,
    formData,
    height,
    hooks,
    initialValues,
    queryData,
    queriesData,
    width
  }));
};
