// @superset-ui/core/lib/chart/models/ChartProps.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _reselect = require("reselect");

var _ = require("../..");

const DEFAULT_WIDTH = 800;
const DEFAULT_HEIGHT = 600;

class ChartProps {
  constructor(config = {}) {
    const {
      annotationData = {},
      datasource = {},
      formData = {},
      hooks = {},
      initialValues = {},
      queryData = {},
      queriesData = [],
      width = DEFAULT_WIDTH,
      height = DEFAULT_HEIGHT
    } = config;
    this.width = width;
    this.height = height;
    this.annotationData = annotationData;
    this.datasource = (0, _.convertKeysToCamelCase)(datasource);
    this.rawDatasource = datasource;
    this.formData = (0, _.convertKeysToCamelCase)(formData);
    this.rawFormData = formData;
    this.hooks = hooks;
    this.initialValues = initialValues;
    this.queryData = queryData;
    this.queriesData = queriesData;
  }

} // eslint-disable-next-line func-name-matching


exports.default = ChartProps;

ChartProps.createSelector = function create() {
  return (0, _reselect.createSelector)(input => input.annotationData, input => input.datasource, input => input.formData, input => input.height, input => input.hooks, input => input.initialValues, input => input.queryData, input => input.queriesData, input => input.width, (annotationData, datasource, formData, height, hooks, initialValues, queryData, queriesData, width) => new ChartProps({
    annotationData,
    datasource,
    formData,
    height,
    hooks,
    initialValues,
    queryData,
    queriesData,
    width
  }));
};
